File: development-kit/pkg/databases/relational/repository/account/account.go
// Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package account

import (
	SQL "github.com/ZupIT/horusec/development-kit/pkg/databases/relational"
	authEntities "github.com/ZupIT/horusec/development-kit/pkg/entities/auth"
	"github.com/google/uuid"
)

type IAccount interface {
	Create(account *authEntities.Account) error
	GetByAccountID(accountID uuid.UUID) (*authEntities.Account, error)
	GetByEmail(email string) (*authEntities.Account, error)
	Update(account *authEntities.Account) error
	GetByUsername(username string) (*authEntities.Account, error)
	DeleteAccount(accountID uuid.UUID) error
}

type Account struct {
	databaseRead  SQL.InterfaceRead
	databaseWrite SQL.InterfaceWrite
}

func NewAccountRepository(databaseRead SQL.InterfaceRead, databaseWrite SQL.InterfaceWrite) IAccount {
	return &Account{
		databaseRead:  databaseRead,
		databaseWrite: databaseWrite,
	}
}

func (a *Account) Create(account *authEntities.Account) error {
	return a.databaseWrite.Create(account, account.GetTable()).GetError()
}

func (a *Account) GetByAccountID(accountID uuid.UUID) (*authEntities.Account, error) {
	account := &authEntities.Account{}
	filter := a.databaseRead.SetFilter(map[string]interface{}{"account_id": accountID})
	result := a.databaseRead.Find(account, filter, account.GetTable())
	return account, result.GetError()
}

func (a *Account) GetByEmail(email string) (*authEntities.Account, error) {
	account := &authEntities.Account{}
	filter := a.databaseRead.SetFilter(map[string]interface{}{"email": email})
	result := a.databaseRead.Find(account, filter, account.GetTable())
	return account, result.GetError()
}

func (a *Account) Update(account *authEntities.Account) error {
	account.SetUpdatedAt()
	return a.databaseWrite.Update(account.ToUpdateMap(), map[string]interface{}{"account_id": account.AccountID},
		account.GetTable()).GetError()
}

func (a *Account) GetByUsername(username string) (*authEntities.Account, error) {
	account := &authEntities.Account{}
	filter := a.databaseRead.SetFilter(map[string]interface{}{"username": username})
	result := a.databaseRead.Find(account, filter, account.GetTable())
	return account, result.GetError()
}

func (a *Account) DeleteAccount(accountID uuid.UUID) error {
	return a.databaseWrite.Delete(map[string]interface{}{"account_id": accountID}, "accounts").GetError()
}

File: horusec-manager/src/services/account.ts
/**
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import http from 'config/axios/default';
import axios from 'axios';
import { SERVICE_AUTH } from '../config/endpoints';
import {
  setCurrentUser,
  clearCurrentUser,
} from 'helpers/localStorage/currentUser';
import { AxiosResponse, AxiosError } from 'axios';
import { User } from 'helpers/interfaces/User';
import { getAccessToken, getRefreshToken } from 'helpers/localStorage/tokens';
import { LoginParams } from 'helpers/interfaces/LoginParams';

const login = (params: LoginParams) => {
  return http.post(`${SERVICE_AUTH}/api/auth/authenticate`, params);
};

const logout = () => http.post(`${SERVICE_AUTH}/api/account/logout`);

const createAccount = (username: string, password: string, email: string) => {
  return http.post(`${SERVICE_AUTH}/api/account/create-account`, {
    username,
    email,
    password,
  });
};

const update = (username: string, email: string) => {
  return http.patch(`${SERVICE_AUTH}/api/account/update`, {
    username,
    email,
  });
};

const deleteAccount = () => {
  return http.delete(`${SERVICE_AUTH}/api/account/delete`);
};

const createAccountFromKeycloak = (accessToken: string) => {
  return http.post(`${SERVICE_AUTH}/api/account/create-account-from-keycloak`, {
    accessToken,
  });
};

const sendCode = (email: string) => {
  return http.post(`${SERVICE_AUTH}/api/account/send-code`, { email });
};

const validateCode = (email: string, code: string) => {
  return http.post(`${SERVICE_AUTH}/api/account/validate-code`, {
    email,
    code,
  });
};

const changePassword = (token: string, password: string) => {
  return http.post(`${SERVICE_AUTH}/api/account/change-password`, password, {
    headers: {
      'Content-Type': 'text/plain',
      Authorization: `Bearer ${token}`,
    },
  });
};

const updatePassword = (password: string) => {
  return http.post(`${SERVICE_AUTH}/api/account/change-password`, password, {
    headers: {
      'Content-Type': 'text/plain',
    },
  });
};

const verifyUniqueUsernameEmail = (email: string, username: string) => {
  return http.post(`${SERVICE_AUTH}/api/account/verify-already-used`, {
    email,
    username,
  });
};

const callRenewToken = async (): Promise<User | AxiosError> => {
  const accessToken = getAccessToken();
  const refreshToken = getRefreshToken();

  const handleLogout = () => {
    clearCurrentUser();
    window.location.replace('/auth');
  };

  if (refreshToken) {
    return new Promise((resolve, reject) => {
      axios
        .post(`${SERVICE_AUTH}/api/account/renew-token`, refreshToken, {
          headers: {
            Authorization: `Bearer ${accessToken}`,
            'Content-type': 'text/plain',
          },
        })
        .then((result: AxiosResponse) => {
          const user = result.data?.content as User;

          if (user) {
            setCurrentUser(user);
          }

          resolve(user);
        })
        .catch((err: AxiosError) => {
          reject(err);
          handleLogout();
        });
    });
  } else {
    handleLogout();
  }
};

const getHorusecConfig = () => {
  return axios.get(`${SERVICE_AUTH}/api/auth/config`);
};

export default {
  login,
  logout,
  createAccount,
  sendCode,
  validateCode,
  changePassword,
  callRenewToken,
  verifyUniqueUsernameEmail,
  getHorusecConfig,
  createAccountFromKeycloak,
  update,
  deleteAccount,
  updatePassword,
};
